<h2>Знакомство</h2>

Меня зовут Клышинский Эдуард Станиславович. <br>
В течение этого года мы будем заниматься изучением возможностей языка Питон, а также библиотек для него.<br>
Со мной можно связаться по почте eklyshinsky@hse.ru<br>

<h2>Морфологический анализ</h2>

На начальных этапах обработки текста проводится два этапа анализа: <b>графематический</b> (выделение предложений и слов) и <b>морфологический</b> (определение начальной формы слова, его части речи и грамматических параметров). Этап синтаксического анализа мы сейчас разбирать не будем, так как его информация требуется не всегда.<br>
Задачей графематического анализа является разделение текста на составные части - врезки, абзацы, предложения, слова. В таких задачах как машинный перевод, точность данного этапа может существенно влиять на точность получаемых результатов. Например, точка, используемая для сокращений, может быть воспринята как конец предложения, что полность разорвет его семантику.<br>
Задачей морфологического анализа является определение начальной формы слова, его части речи и грамматических параметров. В некоторых случаях от слова требуется только начальная форма, в других - только начальная форма и часть речи.<br>
Существует два больших подхода к морфологическому анализу: <b>стемминг</b> и <b>поиск по словарю</b>. Для проведения стемминга оставляется справочник всех окончаний для данного языка. Для пришедшего слова проверяется его окончание и по нему делается прогноз начальной формы и части речи.<br>
Например, мы создаем справочник, в котором записываем все окончания прилагательных: <i>-ому, -ему, -ой, -ая, -ий, -ый, ...</i> Теперь все слова, которые имеют такое окончание будут считаться прилагаельными: <i>синий, циклический, красного, больному</i>. Заодно прилагательными будут считаться причастия (<i>делающий, строившему</i>) и местоимения (<i>мой, твой, твоему</i>). Также не понятно что делать со словами, имеющими пустое окончание. Отдельную проблему составляют такие слова, как <i>стекло, больной, вина</i>, которые могут разбираться несколькими вариантами (это явление называется <b>омонимией</b>). Помимо этого, стеммер может просто откусывать окончания, оставляя лишь псевдооснову.<br>
Большинство проблем здесь решается, но точность работы бессловарных стеммеров находится на уровне 80%. Чтобы повысить точность испольуют морфологический анализ со словарем. Разработчики составляют словарь слов, встретившихся в текстах (<a href="http://opencorpora.org/dict.php">здесь</a> можно найти пример такого словаря). Теперь каждое слово будет искаться в словаре и не предсказываться, а выдаваться точно. Для слов, отсутствующих в словаре, может применяться предсказание, пообное работе стеммера.<br>
Посмотрим как работает словарная морфология на примере системы <a href="https://pymorphy2.readthedocs.io/en/latest/">pymorphy2</a>.

In [10]:
import pymorphy2 # Импортируем морфологический анализатор.

In [96]:
type(morph.cyr2lat)

method

In [11]:
morph=pymorphy2.MorphAnalyzer() # Создает объект морфоанализатора и загружет словарь.
wordform=morph.parse('стекло')  # Проведем анализ слова "стекло".
for w in wordform: # Возвращается список вариантов разбора.
    print(w)# Посмотрим на полученный результат.

Parse(word='стекло', tag=OpencorporaTag('NOUN,inan,neut sing,nomn'), normal_form='стекло', score=0.75, methods_stack=((<DictionaryAnalyzer>, 'стекло', 545, 0),))
Parse(word='стекло', tag=OpencorporaTag('NOUN,inan,neut sing,accs'), normal_form='стекло', score=0.1875, methods_stack=((<DictionaryAnalyzer>, 'стекло', 545, 3),))
Parse(word='стекло', tag=OpencorporaTag('VERB,perf,intr neut,sing,past,indc'), normal_form='стечь', score=0.0625, methods_stack=((<DictionaryAnalyzer>, 'стекло', 968, 3),))


In [82]:
morph=pymorphy2.MorphAnalyzer() # Создает объект морфоанализатора и загружет словарь.
wordform=morph.parse('стекло')  # Проведем анализ слова "стекло".
for w in wordform: # Возвращается список вариантов разбора.
    if 'NOUN' in w.tag:
        print(w.normal_form)# Посмотрим на полученный результат.

стекло
стекло


In [85]:
morph.parse('делавшихся')

[Parse(word='делавшихся', tag=OpencorporaTag('PRTF,impf,intr,past,actv plur,gent'), normal_form='делаться', score=0.3333333333333333, methods_stack=((<DictionaryAnalyzer>, 'делавшихся', 224, 61),)),
 Parse(word='делавшихся', tag=OpencorporaTag('PRTF,impf,intr,past,actv anim,plur,accs'), normal_form='делаться', score=0.3333333333333333, methods_stack=((<DictionaryAnalyzer>, 'делавшихся', 224, 63),)),
 Parse(word='делавшихся', tag=OpencorporaTag('PRTF,impf,intr,past,actv plur,loct'), normal_form='делаться', score=0.3333333333333333, methods_stack=((<DictionaryAnalyzer>, 'делавшихся', 224, 66),))]

Как видно из вывода, слово "стекло" может быть неодушевленным существительным среднего рода, единственного числа, именительного падежа <i>tag=OpencorporaTag('NOUN,inan,neut sing,nomn')</i>, аналогично, но в винительном падеже (<i>'NOUN,inan,neut sing,accs'</i>), и глаголом <i>'VERB,perf,intr neut,sing,past,indc'</i>. При этом в первой форме оно встречается в 75% случаев (<i>score=0.75</i>), во второй в 18,75% случаев (<i>score=0.1875</i>), а как глагол - лишь в 6,25% (<i>score=0.0625</i>). Самым простым видом борьбы с омонимией является выбор нулевого элемента из списка, возвращенного морфологическим анализом. Такой подход дает около 90% точности при выборе начальной формы и до 80% если мы обращаем внимание на грамматические параметры.<br>

Вместо Pymorphy можно использовать PyMystem. Его плюсом является тот факт, что он сам проводит графематический анализ и снимает омонимию. Используя функцию lemmatize можно получить набор начальных форм слов. Используя функцию analyze можно получить полную информацию о словах.

In [13]:
import pymystem3

In [84]:
print(mystem.analyze('Системы делавшегося проектирования'))

[{'analysis': [{'lex': 'система', 'wt': 1, 'gr': 'S,жен,неод=(вин,мн|род,ед|им,мн)'}], 'text': 'Системы'}, {'text': ' '}, {'analysis': [{'lex': 'делаться', 'wt': 1, 'gr': 'V,несов,нп=(прош,вин,ед,прич,полн,муж,действ,од|прош,род,ед,прич,полн,муж,действ|прош,род,ед,прич,полн,сред,действ)'}], 'text': 'делавшегося'}, {'text': ' '}, {'analysis': [{'lex': 'проектирование', 'wt': 1, 'gr': 'S,сред,неод=(вин,мн|род,ед|им,мн)'}], 'text': 'проектирования'}, {'text': '\n'}]


In [14]:
mystem=pymystem3.Mystem()
print(mystem.lemmatize('эти типы стали есть в цеху'))
print()
print(mystem.analyze('эти типы стали есть в цеху'))

['этот', ' ', 'тип', ' ', 'становиться', ' ', 'есть', ' ', 'в', ' ', 'цех', '\n']

[{'analysis': [{'lex': 'этот', 'wt': 1, 'gr': 'APRO=(им,мн|вин,мн,неод)'}], 'text': 'эти'}, {'text': ' '}, {'analysis': [{'lex': 'тип', 'wt': 0.8700298667, 'gr': 'S,муж,неод=(вин,мн|им,мн)'}], 'text': 'типы'}, {'text': ' '}, {'analysis': [{'lex': 'становиться', 'wt': 0.9821285009, 'gr': 'V,нп=прош,мн,изъяв,сов'}], 'text': 'стали'}, {'text': ' '}, {'analysis': [{'lex': 'есть', 'wt': 0.04922361672, 'gr': 'V,несов,пе=инф'}], 'text': 'есть'}, {'text': ' '}, {'analysis': [{'lex': 'в', 'wt': 0.9999917746, 'gr': 'PR='}], 'text': 'в'}, {'text': ' '}, {'analysis': [{'lex': 'цех', 'wt': 1, 'gr': 'S,муж,неод=(дат,ед|местн,ед)'}], 'text': 'цеху'}, {'text': '\n'}]


Функция `lemmatize` делит текст на слова и знаки препинания, а затем возвращает для них только начальную форму.

Функция `analyze` возвращает не только начальную форму, но и всю информацию о слове, как это делал перед этим Pymorphy. 

Основным отличием является то, что Mystem снимает омонимию. Как видно из примера, делает он это не всегда корректно, но нам не придется думать о том, какое вариант разбора следует взять.

Еще одна библиотека - NLTK. По сравнению с двумя предыдущими библиотеками она обладает более широкой функциональностью и изначально писалась для работы с разными языками.

In [1]:
import nltk # Иностранный морфологический анализатор.

Перед началом использования необходимо загрузить необходимые библиотеки или корпуса.

In [ ]:
nltk.download() # По дороге будут появляться поле ввода. Грузит всё из Сети.

In [ ]:
nltk.download('averaged_perceprton_tagger') # Сразу грузит что попросили.

Функция `word_tokenize` возвращает начальные формы слов. 

Функция `pos_tag` возвращает список начальных форм и их частей речи.

In [3]:
tokens = nltk.word_tokenize('Эти типы стали есть в цеху') # Токенизация.
bi_tokens = list(nltk.bigrams(tokens))
tokens, bi_tokens

(['Эти', 'типы', 'стали', 'есть', 'в', 'цеху'],
 [('Эти', 'типы'),
  ('типы', 'стали'),
  ('стали', 'есть'),
  ('есть', 'в'),
  ('в', 'цеху')])

In [4]:
pos = nltk.pos_tag(tokens) # Частеречная разметка.
bi_pos = list(nltk.bigrams(pos))
pos, bi_pos

([('Эти', 'JJ'),
  ('типы', 'NNP'),
  ('стали', 'NNP'),
  ('есть', 'NNP'),
  ('в', 'NNP'),
  ('цеху', 'NN')],
 [(('Эти', 'JJ'), ('типы', 'NNP')),
  (('типы', 'NNP'), ('стали', 'NNP')),
  (('стали', 'NNP'), ('есть', 'NNP')),
  (('есть', 'NNP'), ('в', 'NNP')),
  (('в', 'NNP'), ('цеху', 'NN'))])

У NLTK заведен список стоп-слов, которые лучше фильтровать при анализе текстов. Но их не очень много. Зато самые мешающиеся.

In [5]:
# Оставим только те слова, которых нет в списке стоп-слов.
filtered_words = [token for token in tokens if token not in nltk.corpus.stopwords.words('russian')]
print('всего русских стоп-слов', len(nltk.corpus.stopwords.words('russian')))
filtered_words

всего русских стоп-слов 151


['Эти', 'типы', 'стали', 'цеху']

Ниже приведены примеры функций морфологического анализа текста для разных библиотек.

In [8]:
import re # Подгружаем библиотеку с регулярными выражениями.

In [12]:
# Pymorphy
def normalizePymorphy(morph, text):
    tokens = re.findall('[A-Za-zА-Яа-яЁё]+\-[A-Za-zА-Яа-яЁё]+|[A-Za-zА-Яа-яЁё]+', text)
    words = []
    for t in tokens:
        pv = morph.parse(t)
        words.append(pv[0].normal_form + '_' + str(pv[0].tag.POS)) # Берем наиболее вероятную форму.
    return words    
        
# Обратите внимание, что про иностранные слова словарь ничего не знает.
normalizePymorphy(morph, "Милая мама мыла белую раму.")

['милый_ADJF', 'мама_NOUN', 'мыло_NOUN', 'белый_ADJF', 'рам_NOUN']

In [15]:
# PyMystem
def normalizePymystem(mystem, text):
    tokens = mystem.analyze(text)
    words = []
    for t in tokens:
        if 'analysis' in t.keys():
            if t['analysis'] != []:
                words.append(t['analysis'][0]['lex']+'_'+t['analysis'][0]['gr'][0])
            else:
                words.append(t['text']+'_'+'U')
    return words    
        
# Не все считают, что причастие всегда выступает в роли глагола, но иногда так значительно проще.
normalizePymystem(mystem, "Милая мама мыла белую раму.")

['милый_A', 'мама_S', 'мыло_S', 'белый_A', 'рама_S']

In [16]:
# NLTK
def normalizeNLTK(text):
    tokens = nltk.pos_tag(nltk.word_tokenize(text))
    words = []
    for t in tokens:
        if t[0] != t[1]:
            words.append(t[0]+'_'+t[1])
    return words    
        
# А вот здесь с частеречной разметкой всё плохо, а параметров нет вовсе.
normalizeNLTK("Милая мама мыла белую раму.")

['Милая_JJ', 'мама_NNP', 'мыла_NNP', 'белую_NNP', 'раму_NNP']

Но мы будем использовать pymorphy, так как он немного пошустрее.

Теперь посмотрим с какой скоростью работают эти анализаторы.

In [88]:
re.findall("([А-ЯЁа-яё]+(-[А-ЯЁа-яё]+)*)", textWP)[:10]

[('Марсианин', ''),
 ('Игра', ''),
 ('слов', ''),
 ('англ', ''),
 ('возможность', ''),
 ('прим', ''),
 ('пер', '')]

In [93]:
textWP[:800]

"\n            Лев Николаевич Толстой. Война и мир. Том 1\n\n             * ЧАСТЬ ПЕРВАЯ. * \n\n\n          \n          \n\n            I.\n\n\n          \n               -- Еh bien, mon prince. Gênes et Lucques ne sont plus que des apanages,\n          des поместья, de la famille Buonaparte.  Non, je  vous préviens, que si vous\n          ne  me dites pas, que nous avons la guerre, si vous vous permettez encore de\n          pallier  toutes les infamies, toutes les  atrocités  de cet  Antichrist  (ma\n          parole, j'y  crois) -- je  ne  vous  connais plus, vous n'êtes plus mon ami,\n          vous n'êtes  plus  мой  верный  раб,  comme  vous  dites.  [1]  Ну,\n          здравствуйте, здравствуйте.  Je vois  que  je  vous fais  peur, [2]\n          садитесь и рассказывайте.\n               Так говорила в и"

In [92]:
with open("data/war_and_peace.txt") as fil:
    textWP=fil.read()
# Выделяем все слова написанные русской кириллицей.
words=[w[0] for w in re.findall("([А-ЯЁа-яё]+(-[А-ЯЁа-яё]+)*)", textWP)]
newtext=" ".join(words)

In [18]:
print(len(words)) # Вся "Война и мир" занимает примерно 450 000 слов.

445508


In [22]:
%%time
iniMorphy=[]
for w in words:
    r=morph.parse(w)
    iniMorphy.append(r[0].normal_form) # Берем только начальные формы слов, остальное нам не надо.

CPU times: user 8.34 s, sys: 0 ns, total: 8.34 s
Wall time: 8.4 s


Теперь посмотрим на скорость работы PyMyStem.

In [23]:
%%time
iniStem=mystem.lemmatize(newtext)

CPU times: user 2 s, sys: 151 ms, total: 2.15 s
Wall time: 20.8 s


In [24]:
%%time
iniStem=mystem.analyze(newtext)

CPU times: user 1.59 s, sys: 72.5 ms, total: 1.67 s
Wall time: 19.6 s


<h2>Борьба за производительность</h2>

Примерно 50 000 слов в секунду для PyMorphy и 25 000 (со снятием омонимии) для MyStem.<br>
Хорошо, но хочется быстрее.<br>
Давайте посмотрим на статистику встречаемости слов.

In [44]:
from collections import Counter # Импортируем счетчик из стандартных библиотек.

In [45]:
frqs=Counter(words) # Считаем сколько раз встречается каждое слово, ...
frqsSorted=sorted(frqs.items(), key=lambda x: x[1], reverse=True) # ... сортируем, ...
for x in frqsSorted[:10]: # ... и выводим 10 самых частоных слов.
    print(x[0], ":", frqs[x[0]])

и : 20324
в : 10202
не : 8434
что : 7298
на : 6441
он : 5986
с : 5748
его : 3864
как : 3681
к : 3408


In [46]:
print(sum([x[1] for x in frqsSorted[:10]])/len(words))
print(sum([x[1] for x in frqsSorted[:100]])/len(words))

0.16921357192238973
0.3676791438088654


Сто самых частотных словоформ занимает примерно треть текста!<br>
Возможно, если мы сумеем кешировать результаты работы морфологического анализатора, он начнет работать быстрее. Используем для этого обычный питоновский словарь.

In [94]:
%%time
iniMorphy2=[]
cash={} # Словарь для кеширования.
for w in words:
    if w in cash.keys(): # Если слово было закешировано, возьмем его из словаря.
        iniMorphy2.append(cash[w])
    else: # В противном случае проведем морфологический анализ.
        r=morph.parse(w)
        iniMorphy2.append(r[0].normal_form)
        cash[w]=r[0].normal_form 
    # Вообще-то, можно было кешировать все результаты. Но нам же нужна только наиболее вероятная начальная форма.

CPU times: user 1.46 s, sys: 7.55 ms, total: 1.46 s
Wall time: 1.46 s


Ускорение почти в 6,5 раз!<br>
Имеет смысл сделать из этого какую-то удобную обертку, чтобы повторно использовать в своих дальнейших разработках.

<h2>Создание и использование классов</h2>

Создадим свой собственный класс.<br>
Класс - это тип, определенный пользователем (программистом). Класс содержит в себе как данные, так и функции, которые работают с этими данными.<br>
Так как это тип, то можно создавать переменные этого типа. Каждая переменная будет хранить и обрабатывать свой набор данных.<br>
В каждую функцию класса обязательно передается переменная, которая обычно называется self. Эта переменная содержит в себе объект, для которого производится вызов функции. Мы можем обращаться к свойствам данного объекта, используя или модифицируя тем самым этот объект.

In [100]:
# Ключевое слово class после которого идет название нашего класса.
class FasterMorphology:
    
    def __init__(self): # Функция инициализации объекта после его создания.
        # Создаем новую морфологию в каждом объекте. 
        # А вдруг мы будем потом работать с разными языками? У каждого объекта должна быть своя.
        self.morpho=pymorphy2.MorphAnalyzer() 
        self.cash={} # Создаем словарь для кеширования.
        
    def analyzeWords(self, words) -> list:
        """ Multiline comment after a function comes to its documentation.
            The function's documentation is accessible by Shift+Tab.
            
            This function analyses a list of tokens using pymorphy2 and hashes result for faster processing.
            
            words - list of words.
        """
        res=[]
        for w in words:
            if w in self.cash:
                res.append(self.cash[w])
            else:
                r=self.morpho.parse(w)[0].normal_form
                res.append(r)
                self.cash[w]=r
        return res

Для создания объекта необходимо вызвать функцию с тем же именем, что и имя его типа.

In [101]:
fast=FasterMorphology()

Теперь у нас есть переменная нужного типа и мы можем обращаться к ее полям, а также вызывать ее методы.

In [102]:
%%time
res=fast.analyzeWords(words) # Нажмите Shift+Tabчтобы посмотреть документацию.

CPU times: user 1.31 s, sys: 14.7 ms, total: 1.32 s
Wall time: 1.33 s


In [53]:
res[:10]

['левый',
 'николай',
 'толстой',
 'война',
 'и',
 'мир',
 'тот',
 'часть',
 'один',
 'е']

In [54]:
print(list(fast.cash.keys())[:10])
print(len(fast.cash.keys()))

['Лев', 'Николаевич', 'Толстой', 'Война', 'и', 'мир', 'Том', 'ЧАСТЬ', 'ПЕРВАЯ', 'Е']
52659


А теперь еще раз то же самое, но с заполненным кешем.

In [105]:
%%time
res=fast.analyzeWords(words)

CPU times: user 140 ms, sys: 4.03 ms, total: 144 ms
Wall time: 143 ms


Ускорение в 10 раз!<br>
Правда, заодно мы выяснили, что слова надо приводить к единому написанию, устраняя, например, заглавные буквы. Повторим эксперимент без них.

In [106]:
words2=[w[0].lower() for w in re.findall("([А-ЯЁа-яё]+(-[А-ЯЁа-яё]+)*)", textWP)]

In [107]:
fast2=FasterMorphology()

In [108]:
%%time
res=fast2.analyzeWords(words2)

CPU times: user 1.26 s, sys: 7.41 ms, total: 1.27 s
Wall time: 1.27 s


<h2>Векторизация текстов</h2>

Теперь напишем класс, отвечающий за векторизацию текста.

In [38]:
import math
import numpy as np

In [33]:
# Другое имя класса, так как он обладает несколько иной функциональностью.
class FasterMorphology2:
    """ Класс для быстрого морфологического анализа текстов и их векторизации.
    """
    
    def __init__(self): # Функция инициализации объекта после его создания.
        self.morpho=pymorphy2.MorphAnalyzer()
        self.cash={}
        self.dictionary={} # Добавим словарь для запоминания, на каком месте вектора находится какая начальная форма.
        
    def analyzeWords(self, words):
        """ Проводит морфологический анализ списка токенов words.
            Возвращает список начальных форм слов.
        """
        res=[]
        for w in words:
            if w in self.cash: # Сперва ищем очередное слово в кеше.
                res.append(self.cash[w])
            else: # Если его там нет, проводим морфологический анализ и кешируем.
                r=self.morpho.parse(w)[0].normal_form
                res.append(r)
                self.cash[w]=r
                if r not in self.dictionary: # Также для каждой начальной формы запоминаем ее позицию в векторе.
                    self.dictionary[r]=len(self.dictionary)+1
        return res
    
    def analyzeText(self, text):
        """ Проводит морфологический анализ строки с текстом text. 
            Выделяет из нее слова, написанные русской кириллицей.
            Возвращает список начальных форм слов.
        """
        words=[w[0] for w in re.findall("([А-ЯЁа-яё]+(-[А-ЯЁа-яё]+)*)", textWP)]
        return self.analyzeWords(words)
        
    # Вообще-то тоже самое умеет Counter, но ему надо сперва привести слова к начальной форме.
    def vectorizeAsDict(self, words):
        """ Возвращает векторное разреженное представление текста в виде словаря.
            Текст передается как список токенов words.
            Вместо позиции для индексации используется само слово.
            Возвращает словарь с начальными формами в ключах и частотами этих форм.
        """
        vct = {}
        res = []
        for w in words: # Для каждого слова прповодим анализ.
            if w in self.cash:
                vct[self.cash[w]]=vct.get(self.cash[w], 0)+1 # Считаем частоты слов.
            else:
                r=self.morpho.parse(w)[0].normal_form
                res.append(r)
                self.cash[w]=r
                vct[r]=vct.get(r, 0)+1
                if r not in self.dictionary:
                    self.dictionary[r]=len(self.dictionary)
        return vct
    
    def clearDict(self):
        """ Очищает словарь. Вдруг надо пересчитать так как изменилась размерность пространства.
        """
        self.dictionary={}
    
    def formDict(self, texts):
        """ Сформировать словарь по тексту не формируя разметку текста.
        """
        for text in texts:
            for word in text:
                if word not in self.cash:
                    r=self.morpho.parse(w)[0].normal_form
                    self.cash[word]=r
                    if r not in self.dictionary:
                        self.dictionary[r]=len(self.dictionary)
    
    def vectorizeAsList(self, words):
        """ Возвращает векторное представление текста в виде плотного списка (включает нули).
            Текст передается как список токенов words.
            Позиция каждого слова в векторе определяется числом, хранимым в dictionary.
            Возвращает словарь с начальными формами в ключах и частотами этих форм.
        """
        # Сперва обновляем dictionary.
        for word in words:
            if word not in self.cash:
                r=self.morpho.parse(w)[0].normal_form
                self.cash[word]=r
                if r not in self.dictionary:
                    self.dictionary[r]=len(self.dictionary.keys())
        # Теперь, когда все слова есть в кеше и словаре и известен размер вектора, можно приступать к векторизации.
        vct=[0 for _ in self.dictionary]
        for word in words:
            vct[self.dictionary[self.cash[word]]]+=1
        return vct
    
    def vectorizeAsList2(self, words):
        """ Возвращает векторное представление текста в виде плотного списка (включает нули).
            Текст передается как список токенов words. В вектор включаются только слова, находящиес в словаре.
            Позиция каждого слова в векторе определяется числом, хранимым в dictionary.
            Возвращает словарь с начальными формами в ключах и частотами этих форм.
        """
        vct=[0 for _ in self.dictionary]
        for word in words:
            if word in self.cash:
                vct[self.dictionary[self.cash[word]]]+=1
        return vct

    def vectorizeAsArray(self, words):
        """ Возвращает векторное представление текста в виде плотного массива (включает нули).
            Текст передается как список токенов words.
            Позиция каждого слова в векторе определяется числом, хранимым в dictionary.
            Возвращает словарь с начальными формами в ключах и частотами этих форм.
        """
        # Сперва обновляем dictionary.
        for word in words:
            if word not in self.cash:
                r=self.morpho.parse(w)[0].normal_form
                self.cash[word]=r
                if r not in self.dictionary:
                    self.dictionary[r]=len(self.dictionary.keys())
        # Теперь, когда все слова есть в кеше и словаре и известен размер вектора, можно приступать к векторизации.
        vct=np.zeros((len(self.dictionary)))
        for word in words:
            vct[self.dictionary[self.cash[word]]]+=1
        return vct

    # Здесь мы заложили проблему. Функция не умеет считать расстояние между p.array.
    def cosineSimilarity(self, a, b):
        """ Функция расчета косинусной меры сходства между двумя векторными представлениями текста.
            Работает по-разному в зависимости от представления вектора.
        """
        if type(a)!=type(b): # Тип векторов должен совпадать.
            return None
        if isinstance(a, list): # Если это списки, значит это плотное представление вектора.
            if len(a)==0 or len(b)==0 or len(a)!=len(b): # Длины векторов в этом случае должны совпадать.
                return 0
            sumab=sum([a[na]*b[na] for na in range(len(a))])
            suma2=sum([a[na]*a[na] for na in range(len(a))])
            sumb2=sum([b[na]*b[na] for na in range(len(a))])
            return sumab/math.sqrt(suma2*sumb2)        
        elif isinstance(a, dict): # Разреженное представление вектора - хранятся только ненулевые значения.
            if len(a.keys())==0 or len(b.keys())==0: # Вектора должны хранить хоть что-то.
                return 0
            sumab=sum([a[na]*b[na] for na in set(a.keys())&set(b.keys())])
#            sumab=sum([a[na]*b[na] for na in a.keys() if na in b.keys()])
            suma2=sum([a[na]*a[na] for na in a.keys()])
            sumb2=sum([b[nb]*b[nb] for nb in b.keys()])
            return sumab/math.sqrt(suma2*sumb2)  
        return 0
    
    def JaccardCoefficient(self, a, b):
        """ Коэффициент Жаккара - отношение количества слов, встречающихся в обоих текстах к объединению лексики.
        """
        if type(a)!=type(b): # Тип векторов должен совпадать.
            return None
        if isinstance(a, list): # Если это списки, значит это плотное представление вектора.
            if len(a)==0 or len(b)==0 or len(a)!=len(b): # Длины векторов в этом случае должны совпадать.
                return 0
            union=len(a)-[aa*bb for aa, bb in zip(a, b)].count(0)
            intersection=len(a)-[aa+bb for aa, bb in zip(a, b)].count(0)
            return union/intersection
        elif isinstance(a, dict): # Разреженное представление вектора - хранятся только ненулевые значения.
            if len(a.keys())==0 or len(b.keys())==0: # Вектора должны хранить хоть что-то.
                return 0
            return len(set(a.keys())&set(b.keys()))/len(set(a.keys())|set(b.keys()))
        return 0
        

Для определения меры сходства двух текстов используется косинусная мера сходства, рассчитываемая по следующей формуле: $cos(a,b)=\frac{\sum{a_i * b_i}}{\sqrt {\sum{a_i^2}*\sum{b_i^2}}}$.<br>
Вообще-то, использовать стандартную функцию рассчета косинусной меры сходства из <a href="http://scikit-learn.org/stable/modules/generated/sklearn.metrics.pairwise.cosine_similarity.html">sklearn</a> было бы быстрее. Но мне хотелось показать как работать с разными типами входа.

Посмотрим как работает новая морфология. Возьмем также один из "Севастопольских рассказов" того же автора, чтобы было что использовать при расчете косинусной меры сходства.

In [35]:
faster3=FasterMorphology2()

In [62]:
with open("data/sebastopol.txt") as fil:
    textWP=fil.read()
words3=[w[0].lower() for w in re.findall("([А-ЯЁа-яё]+(-[А-ЯЁа-яё]+)*)", textWP)]
newtext3=" ".join(words3)

Создадим плотное и разреженное представление для двух текстов.

In [63]:
vd1=faster3.vectorizeAsDict(words)
vd2=faster3.vectorizeAsDict(words3)
vl1=faster3.vectorizeAsList(words)
vl2=faster3.vectorizeAsList(words3)

Посмотрим как быстро считается разреженное и плотное предствления.

In [71]:
%time
r1=faster3.cosineSimilarity(vd1, vd2)
%time
r2=faster3.cosineSimilarity(vl1, vl2)
print(r1, r2)

CPU times: user 3 µs, sys: 0 ns, total: 3 µs
Wall time: 5.72 µs
CPU times: user 4 µs, sys: 0 ns, total: 4 µs
Wall time: 8.82 µs
0.9573318479299211 0.9573318479299211


In [72]:
print(vl1.count(0), vl2.count(0), len(vl1))

533 15363 18891


In [73]:
%time
r1=faster3.JaccardCoefficient(vd1, vd2)
%time
r2=faster3.JaccardCoefficient(vl1, vl2)
print(r1, r2)

CPU times: user 8 µs, sys: 0 ns, total: 8 µs
Wall time: 16.5 µs
CPU times: user 4 µs, sys: 1e+03 ns, total: 5 µs
Wall time: 9.06 µs
0.15854110422952727 0.15854110422952727


In [76]:
import sys
print(sys.getsizeof(vd1))
print(sys.getsizeof(vl1))

589928
158200


Разницs по скорости почти нет при разнице словарей 6 раз. И с памятью все тоже не так однозначно.<br>
Видимо, поиск по словарю и хранение строк обходится довольно дорого.<br>
Попробуем заменить слова на их номера.

In [290]:
vdd1={faster3.dictionary[w]: f  for w, f in vd1.items()}
print(sys.getsizeof(vdd1))

589928


Легче не стало. Попробуем с numpy.array.

In [77]:
va1=faster3.vectorizeAsArray(words)
print(sys.getsizeof(va1))

151224


Ну хоть здесь какой-то выигрыш. Мораль - проще разработать свой собственный класс для разреженного хранения на основе общего для всех словаря с индексом и двумя массивами для индекса слов конкретного текста и их частот.<br>
Хорошо, продолжим с косинусной мерой. Попробуем посчитать сходство с "Марсианином" Энди Вейра.

In [78]:
with open("index_split_017.xhtml") as fil: # Грузим главу 15, она побольше.
    textWP=fil.read()
words4=[w[0].lower() for w in re.findall("([А-ЯЁа-яё]+(-[А-ЯЁа-яё]+)*)", textWP)]
newtext4=" ".join(words4)

In [79]:
# Считаем вектора.
vd3=faster3.vectorizeAsDict(words4)
vl3=faster3.vectorizeAsList(words4)

In [84]:
%%time
print(faster3.cosineSimilarity(vd1, vd3))
print(faster3.cosineSimilarity(vd2, vd3))

0.7861664581749179
0.7785426864805731
CPU times: user 11.4 ms, sys: 0 ns, total: 11.4 ms
Wall time: 11.1 ms


In [81]:
%%time
print(faster3.cosineSimilarity(vl1, vl3))
print(faster3.cosineSimilarity(vl2, vl3))

0
0
CPU times: user 306 µs, sys: 20 µs, total: 326 µs
Wall time: 209 µs


Что-то пошло не так. Постараемся понять что именно.

In [178]:
sumab=sum([vl1[na]*vl3[na] for na in range(len(vl3))])
suma2=sum([vl1[na]*vl1[na] for na in range(len(vl3))])
sumb2=sum([vl3[na]*vl3[na] for na in range(len(vl3))])
sumab/math.sqrt(suma2*sumb2) 

IndexError: list index out of range

Ну конечно же! Надо же пересчитать все вектора, а то размерность пространства изменилась! Со словарями такой проблемы не было.

In [82]:
vl1=faster3.vectorizeAsList(words)
vl2=faster3.vectorizeAsList(words3)

In [85]:
%%time
print(faster3.cosineSimilarity(vl1, vl3))
print(faster3.cosineSimilarity(vl2, vl3))

0.7861664581749179
0.7785426864805731
CPU times: user 19.3 ms, sys: 0 ns, total: 19.3 ms
Wall time: 18.8 ms


In [86]:
vl1.count(0), vl2.count(0), vl3.count(0), len(vl1)

(960, 15790, 17715, 19318)

Теперь попробуем построить вектора для всего текста "Марсианина".

In [87]:
words5=[]
for i in range(2, 33):
    with open("data/veyr/index_split_0"+"{:0>2}".format(i)+".xhtml") as fil:
        textWP=fil.read()
    words6=[w[0].lower() for w in re.findall("([А-ЯЁа-яё]+(-[А-ЯЁа-яё]+)*)", textWP)]
    words5+=words6
newtext5=" ".join(words5)

In [88]:
vd4=faster3.vectorizeAsDict(words5)
vl4=faster3.vectorizeAsList(words5)

In [89]:
vl1=faster3.vectorizeAsList(words)
vl2=faster3.vectorizeAsList(words3)
vl3=faster3.vectorizeAsList(words4)

In [90]:
%%time
print(faster3.cosineSimilarity(vd1, vd4))
print(faster3.cosineSimilarity(vd2, vd4))
print(faster3.cosineSimilarity(vd3, vd4))

0.8042099465818117
0.8078103349350544
0.8822254560518593
CPU times: user 22.9 ms, sys: 3.94 ms, total: 26.8 ms
Wall time: 25.4 ms


Да, глава из "Марсианина" больше похожа на всё произведение, чем на Толстого. Но общая мера сходства довольно большая. То есть по-хорошему, речь идет в большой степени об одинаковых вещах.

In [91]:
%%time
print(faster3.cosineSimilarity(vl1, vl4))
print(faster3.cosineSimilarity(vl2, vl4))
print(faster3.cosineSimilarity(vl3, vl4))

0.8042099465818117
0.8078103349350544
0.8822254560518593
CPU times: user 21 ms, sys: 0 ns, total: 21 ms
Wall time: 20.8 ms


In [92]:
print(len(vl1)-vl1.count(0), len(vl1)-vl2.count(0), len(vl1)-vl3.count(0), len(vl1)-vl4.count(0), len(vl1))
print(len(words), len(words5))

18358 3528 1603 8057 22291
445508 90841


In [202]:
with open("war_and_peace3.txt") as fil:
    textWP=fil.read()
words6=[w[0] for w in re.findall("([А-ЯЁа-яё]+(-[А-ЯЁа-яё]+)*)", textWP)]
print(len(words6))
faster4=FasterMorphology2()
vd5=faster4.vectorizeAsDict(words6)
vl5=faster4.vectorizeAsList(words6)
print(len(vl5))

104178
8780


<h2>CountVectorizer и TfidfVectorizer</h2>

На самом деле примерно всё то же самое можно селать при помощи класса CountVectorizer из sklearn.feature_extraction.text. При помощи функции <i>fit\_transform</i> можно получить разреженное представление матрицы частот слов. Основная проблема состоит в том, что индексы в матрице представляют собой индексы в словаре переданных текстов. Сам словарь хранится в свойстве <i>vocabulary\_</i> и умеет возвращать индекс по слову (но не наоборот).

In [66]:
from sklearn.feature_extraction.text import CountVectorizer

In [353]:
counter=CountVectorizer()
# Просим посчитать частоты слов.
res=counter.fit_transform([newtext, newtext3, newtext5])
# Разреженное представление счетчика.
print(res[0][0,:10])
# Можно получить индекс по слову, ...
print(counter.vocabulary_.get('левый'))
# ... но не наоборот.
print(counter.vocabulary_.get(20342))

  (0, 9)	3
  (0, 3)	1
  (0, 4)	1
  (0, 2)	1
  (0, 0)	1
  (0, 7)	3
  (0, 6)	7
  (0, 5)	10
20342
None


Более того, CountVectorizer просто выделяет подстроки и ничего не знает про морфологию (ее можно правильно прикрутить, но это хлопотное занятие). Зато он умеет выделять n-граммы (n слов идущих подряд (или даже букв)). Помимо этого, можно попросить выдать все подстроки, создав анализатор. И можно сказать как выделять подстроки при помощи регулярного выражения.

In [334]:
def getMeaningfullWords(text):
    words=[]
    tokens=re.findall('[А-Яа-яЁё]+\-[А-Яа-яЁё]+|[А-Яа-яЁё]+', text)
    for t in tokens:
        pv=morph.parse(t)
        if pv[0].tag.POS in ['ADJF', 'NOUN', 'VERB', 'PRTF', 'GRND']:
            words.append(pv[0].normal_form)
    return words

lemmaCounter=CountVectorizer(ngram_range=(1,3), token_pattern=r'[А-Яа-яЁё]+\-[А-Яа-яЁё]+|[А-Яа-яЁё]+')

c=[' '.join(getMeaningfullWords(newtext)), ' '.join(getMeaningfullWords(newtext3)), ' '.join(getMeaningfullWords(newtext5))]
analyze = lemmaCounter.build_analyzer()
res1=analyze(c[0])
res2=lemmaCounter.fit_transform(c)

In [335]:
print(res1[:10])

['николай', 'толстой', 'война', 'мир', 'тот', 'часть', 'один', 'е', 'поместье', 'мой']


Теперь попробуем другой показатель для подсчета важности слов в тексте - $TF*IDF$. Здесь $TF$ - Term Frequency, частота термина в документе, а $IDF$ - Inverted Document Frequency, обратная частота термина в коллекции (количество документов, в которых встречается данный термин).

Идея метрики очень проста. Если слово встречается почти во всех документах - его различительная сила очень мала и само слово не является важным. Если слово часто встречается в данном документе, то оно являетсяя важным для него.

Метрика считается на коллекции документов для каждого слова, каждого документа. Для расчета меры можно использовать `TfidfVectorizer`, который работает так же как `CountVectorizer`.

In [44]:
from sklearn.feature_extraction.text import TfidfVectorizer

In [358]:
lemmaCounter=TfidfVectorizer(ngram_range=(1,3), token_pattern=r'[А-Яа-яЁё]+\-[А-Яа-яЁё]+|[А-Яа-яЁё]+')

c=[' '.join(getMeaningfullWords(newtext)), ' '.join(getMeaningfullWords(newtext3)), ' '.join(getMeaningfullWords(newtext5))]
analyze = lemmaCounter.build_analyzer()
res1=analyze(c[0])
res2=lemmaCounter.fit_transform(c)

In [359]:
print(res2[0][0,:10])

  (0, 3)	0.0001129089685194666
  (0, 0)	0.0001129089685194666
  (0, 9)	0.0002258179370389332
  (0, 4)	0.0001129089685194666
  (0, 1)	0.0001129089685194666
  (0, 5)	0.0001129089685194666
  (0, 2)	0.0001129089685194666


Попробуем посмотреть на близость глав "Марсианина" при помощи косинусной меры по частотам слов.

In [30]:
wordsM=[]
for i in range(2, 33):
    with open("data/veyr/index_split_0"+"{:0>2}".format(i)+".xhtml") as fil:
        textWP=fil.read()
    words6=[w[0].lower() for w in re.findall("([А-ЯЁа-яё]+(-[А-ЯЁа-яё]+)*)", textWP)]
    wordsM.append(words6)


In [36]:
m_vects = []

for words in wordsM:
    m_vects.append(faster3.vectorizeAsDict(words))


In [42]:
nearest = -1
txt1 = -1
txt2 = -1
for i, vct1 in enumerate(m_vects):
    for j, vct2 in enumerate(m_vects):
        if vct1 is vct2:
            continue
        cc = faster3.cosineSimilarity(vct1, vct2)
        if cc > nearest:
            nearest = cc
            txt1 = i
            txt2 = j
print(nearest, txt1+2, txt2+2)

0.9424772735909387 21 22


А теперь возьмем косинусную меру сходства по результатам TF*IDF.

In [74]:
textM = []
for i in range(2, 33):
    with open("data/veyr/index_split_0"+"{:0>2}".format(i)+".xhtml") as fil:
        textM.append(fil.read())

In [77]:
lemmaCounter=CountVectorizer(ngram_range=(1,3), token_pattern=r'[А-Яа-яЁё]+\-[А-Яа-яЁё]+|[А-Яа-яЁё]+')
#lemmaCounter=TfidfVectorizer(ngram_range=(1,3), token_pattern=r'[А-Яа-яЁё]+\-[А-Яа-яЁё]+|[А-Яа-яЁё]+')

analyze = lemmaCounter.build_analyzer()
#res1=analyze(c[0])
res2=lemmaCounter.fit_transform(textM)

In [59]:
from sklearn.metrics.pairwise import cosine_similarity

In [80]:
nearest = -1
txt1 = -1
txt2 = -1
for i, vct1 in enumerate(res2):
    for j, vct2 in enumerate(res2):
        if i==j:
            continue
        cc = cosine_similarity(vct1, vct2)
        if cc > nearest:
            nearest = cc
            txt1 = i
            txt2 = j
print(nearest, txt1+2, txt2+2)



[[0.85149324]] 15 21


<h2>Из чего же сделаны классы?</h2>

Теперь посмотрим на объект изнутри. Мы уже обратили внимание, что объект класса надо пересоздать после внесения изменения в класс. Получается, что каждый объект "носит" с собой все свои функции.<br>
Среди прочего, это связано с тем, что Питон работает со ссылками на переменные, а не самими переменными. Переменная в Питоне - это не классический "ящик" в котором хранится значение. Это ссылка на подобный ящик.<br>
Функция <i>dir</i> выдает список всех полей и методов объекта.

In [360]:
dir(faster3)

['JaccardCoefficient',
 '__class__',
 '__delattr__',
 '__dict__',
 '__dir__',
 '__doc__',
 '__eq__',
 '__format__',
 '__ge__',
 '__getattribute__',
 '__gt__',
 '__hash__',
 '__init__',
 '__init_subclass__',
 '__le__',
 '__lt__',
 '__module__',
 '__ne__',
 '__new__',
 '__reduce__',
 '__reduce_ex__',
 '__repr__',
 '__setattr__',
 '__sizeof__',
 '__str__',
 '__subclasshook__',
 '__weakref__',
 'analyzeText',
 'analyzeWords',
 'cash',
 'clearDict',
 'cosineSimilarity',
 'dictionary',
 'formDict',
 'morpho',
 'vectorizeAsArray',
 'vectorizeAsDict',
 'vectorizeAsList',
 'vectorizeAsList2']

А поле \__dict\__ хранит только поля объекта. Но собственно хранит, а не содержит список названий.

In [94]:
faster3.__dict__

{'cash': {'Лев': 'левый',
  'Николаевич': 'николай',
  'Толстой': 'толстой',
  'Война': 'война',
  'и': 'и',
  'мир': 'мир',
  'Том': 'тот',
  'ЧАСТЬ': 'часть',
  'ПЕРВАЯ': 'один',
  'Е': 'е',
  'поместья': 'поместье',
  'мой': 'мой',
  'верный': 'верный',
  'раб': 'раб',
  'Ну': 'ну',
  'здравствуйте': 'здравствовать',
  'садитесь': 'садиться',
  'рассказывайте': 'рассказывать',
  'Так': 'так',
  'говорила': 'говорить',
  'в': 'в',
  'июле': 'июль',
  'года': 'год',
  'известная': 'известный',
  'Анна': 'анна',
  'Павловна': 'павло',
  'Шерер': 'шерер',
  'фрейлина': 'фрейлина',
  'приближенная': 'приблизить',
  'императрицы': 'императрица',
  'Марии': 'мария',
  'Феодоровны': 'феодоровный',
  'встречая': 'встречать',
  'важного': 'важный',
  'чиновного': 'чиновный',
  'князя': 'князь',
  'Василия': 'василий',
  'первого': 'первое',
  'приехавшего': 'приехать',
  'на': 'на',
  'ее': 'она',
  'вечер': 'вечер',
  'кашляла': 'кашлять',
  'несколько': 'несколько',
  'дней': 'день',
  'у':

Теперь добавим объекту несколько новых полей и функций и посмотрим как изменится список (при помощи оператора -).

In [17]:
faster5=FasterMorphology2()
faster4=FasterMorphology2()
faster4.dummy=0

In [18]:
set(dir(faster4))-set(dir(faster5))

{'dummy'}

Допустим, нам вдруг захотелось, чтобы faster4 начал считать Евклидово расстояние. Для этого добавим в объект соответствующую функцию.

In [21]:
def EuclideDistance(self, a, b):
    return math.sqrt(sum([aa*bb for aa, bb in zip(a, b)]))

faster4.EuclidianSimilarity=EuclideDistance

In [25]:
faster4.EuclidianSimilarity([1,2,3], [3,4,5])

TypeError: EuclideDistance() missing 1 required positional argument: 'b'

Что-то опять пошло не так. Оказывается в Питоне функции отличаются от методов класса.

In [26]:
print(faster4.analyzeWords)
print(faster4.EuclidianSimilarity)

<bound method FasterMorphology2.analyzeWords of <__main__.FasterMorphology2 object at 0x7ff2af3d7358>>
<function EuclideDistance at 0x7ff2ad78dae8>


Чтобы привязать метод к отдельному объекту необходимо вызвать специальную функцию.

In [27]:
import types

In [28]:
faster4.EuclidianSimilarity=types.MethodType(EuclideDistance, faster4)

А вот поменять класс целиком довольно просто.

In [31]:
class forTests:

    def __init__(self):
        self.aha=0
        self.uhu=1

test1=forTests()
dir(test1)

def dummyFunc(self):
    return 0

forTests.dummy=dummyFunc

test2=forTests()

print(set(dir(test1))-set(dir(test2)))

set()


Но обратите внимание, метод появился теперь у всех объектов данного класса.